plugins/pulp_rpm/plugins/importers/yum/pulp_solv.py

```diff
@@ -676,20 +681,26 @@ def get_repo_units(self, repo_id):
         )
 
     def get_units_from_solvables(self, solvables):
-        """Map a whole list of solvables back into their Pulp units.
+        """Map a whole list of solvables back into their Pulp units, keyed by the repo they
+        come from.
         """
-        ret = set()
-        for solvables in misc_utils.paginate(solvables):
-            unit_ids = [self.get_unit_id(solvable)[0] for solvable in solvables]
-            iterator = itertools.chain(
-                models.RPM.objects.filter(id__in=unit_ids).only(
-                    *models.RPM.unit_key_fields),
-                models.Modulemd.objects.filter(id__in=unit_ids).only(
-                    *models.Modulemd.unit_key_fields),
-                models.ModulemdDefaults.objects.filter(id__in=unit_ids).only(
-                    *models.ModulemdDefaults.unit_key_fields))
-            for unit in iterator:
-                ret.add(unit)
+        repo_unit_map = collections.defaultdict(set)
+        for solvable in solvables:
+            (unit_id, repo_id) = self.get_unit_id(solvable)
+            repo_unit_map[repo_id].add(unit_id)
+
+        ret = collections.defaultdict(set)
+        for repo, unit_ids in repo_unit_map.items():
+            for paginated_unit_ids in misc_utils.paginate(unit_ids):
```

Comment:
Why do you need to paginate, when all unit_ids are already in memory? Is it because mongo queries below can be too large and exceed BSON limits?